File: adapp-kmp/composeApp/src/androidMain/kotlin/lu/etat/adapp_kmp/views/DeviceView.kt
package lu.etat.adapp_kmp.views

import android.Manifest
import android.content.pm.PackageManager
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import androidx.core.content.ContextCompat
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import lu.etat.adapp_kmp.composables.AccText
import lu.etat.adapp_kmp.composables.DataCard
import lu.etat.adapp_kmp.composables.PageTitle
import lu.etat.adapp_kmp.resources.Labels.DEVICE_INFO_SECTION
import lu.etat.adapp_kmp.resources.Labels.DEVICE_TITLE
import lu.etat.adapp_kmp.resources.Labels.LIVE_METRICS_SECTION
import lu.etat.adapp_kmp.resources.Labels.accelerometerX
import lu.etat.adapp_kmp.resources.Labels.accelerometerY
import lu.etat.adapp_kmp.resources.Labels.accelerometerZ
import lu.etat.adapp_kmp.resources.Labels.altitude
import lu.etat.adapp_kmp.resources.Labels.batteryLevel
import lu.etat.adapp_kmp.resources.Labels.deviceName
import lu.etat.adapp_kmp.resources.Labels.latitude
import lu.etat.adapp_kmp.resources.Labels.longitude
import lu.etat.adapp_kmp.resources.Labels.manufacturer
import lu.etat.adapp_kmp.resources.Labels.model
import lu.etat.adapp_kmp.resources.Labels.osVersion
import lu.etat.adapp_kmp.viewmodels.SharedDeviceViewModel
import org.koin.compose.viewmodel.koinViewModel

@Composable
fun DeviceView() {
    val deviceVM: SharedDeviceViewModel = koinViewModel()

    var hasLocationPermission by remember { mutableStateOf(false) }

    val permissionLauncher = rememberLauncherForActivityResult(
        contract = ActivityResultContracts.RequestMultiplePermissions()
    ) { permissions ->
        hasLocationPermission = permissions[Manifest.permission.ACCESS_FINE_LOCATION] == true
    }

    val context = LocalContext.current
    LaunchedEffect(Unit) {
        val isGranted = ContextCompat.checkSelfPermission(
            context, Manifest.permission.ACCESS_FINE_LOCATION
        ) == PackageManager.PERMISSION_GRANTED

        if (isGranted) {
            hasLocationPermission = true
        } else {
            permissionLauncher.launch(
                arrayOf(
                    Manifest.permission.ACCESS_FINE_LOCATION,
                    Manifest.permission.ACCESS_COARSE_LOCATION
                )
            )
        }
    }

    val accelerometerData by deviceVM.accelerometerData.collectAsStateWithLifecycle()
    val batteryData by deviceVM.batteryData.collectAsStateWithLifecycle()
    val geolocationData by if (hasLocationPermission) {
        deviceVM.geolocationData.collectAsStateWithLifecycle()
    } else {
        remember { mutableStateOf(null) }
    }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .verticalScroll(rememberScrollState())
            .padding(start = 20.dp, top = 40.dp, end = 20.dp, bottom = 40.dp),
        verticalArrangement = Arrangement.spacedBy(15.dp)
    ) {
        PageTitle(DEVICE_TITLE)

        DataCard(DEVICE_INFO_SECTION) {
            AccText(deviceName(deviceVM.deviceDetails.name))
            AccText(manufacturer(deviceVM.deviceDetails.manufacturer))
            AccText(model(deviceVM.deviceDetails.model))
            AccText(osVersion(deviceVM.deviceDetails.osVersion))
        }

        DataCard(LIVE_METRICS_SECTION) {
            AccText(accelerometerX(accelerometerData.x))
            AccText(accelerometerY(accelerometerData.y))
            AccText(accelerometerZ(accelerometerData.z))
            Spacer(modifier = Modifier.height(10.dp))
            AccText(latitude(geolocationData?.latitude))
            AccText(longitude(geolocationData?.longitude))
            AccText(altitude(geolocationData?.altitude))
            Spacer(modifier = Modifier.height(10.dp))
            AccText(batteryLevel(batteryData.level))
        }
    }
}

File: adapp-kmp/iosApp/iosApp/views/DeviceView.swift
import SwiftUI
import Shared

struct DeviceView: View {
    let viewModel: SharedDeviceViewModel

    @State private var accelerometerData: AccelerometerData?
    @State private var geolocationData: GeolocationData?
    @State private var batteryData: BatteryData?

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 15) {
                Title(accessibleText: Labels.shared.DEVICE_TITLE)

                VStack(alignment: .leading, spacing: 10) {
                    AccText(accessibleText: Labels.shared.DEVICE_INFO_SECTION)
                        .font(.system(size: 15, weight: .bold))

                    AccText(accessibleText: Labels.shared.deviceName(deviceName: viewModel.deviceDetails.name))
                    AccText(accessibleText: Labels.shared.manufacturer(manufacturer: viewModel.deviceDetails.manufacturer))
                    AccText(accessibleText: Labels.shared.model(model: viewModel.deviceDetails.model))
                    AccText(accessibleText: Labels.shared.osVersion(osVersion: viewModel.deviceDetails.osVersion))
                }
                .modifier(CardStyle())

                VStack(alignment: .leading, spacing: 10) {
                    AccText(accessibleText: Labels.shared.LIVE_METRICS_SECTION)
                        .font(.system(size: 15, weight: .bold))

                    Group {
                        AccText(accessibleText: Labels.shared.accelerometerX(accelerometerX: accelerometerData?.x ?? 0))
                        AccText(accessibleText: Labels.shared.accelerometerY(accelerometerY: accelerometerData?.y ?? 0))
                        AccText(accessibleText: Labels.shared.accelerometerZ(accelerometerZ: accelerometerData?.z ?? 0))
                            .padding(.bottom, 10)
                    }

                    Group {
                        AccText(accessibleText: Labels.shared.latitude(latitude: KotlinDouble(value: geolocationData?.latitude ?? 0.0)))
                        AccText(accessibleText: Labels.shared.longitude(longitude: KotlinDouble(value: geolocationData?.longitude ?? 0.0)))
                        AccText(accessibleText: Labels.shared.altitude(altitude: KotlinDouble(value: geolocationData?.altitude ?? 0)))
                            .padding(.bottom, 10)
                    }

                    AccText(accessibleText: Labels.shared.batteryLevel(batteryLevel: batteryData?.level ?? 0))
                }
                .modifier(CardStyle())
            }
            .padding(EdgeInsets(top: 40, leading: 20, bottom: 20, trailing: 20))
            .task {
                for await newAccelerometerData in viewModel.accelerometerData {
                    self.accelerometerData = newAccelerometerData
                }
            }
            .task {
                for await newGeolocationData in viewModel.geolocationData {
                    self.geolocationData = newGeolocationData
                }
            }
            .task {
                for await newBatteryData in viewModel.batteryData {
                    self.batteryData = newBatteryData
                }
            }
        }
    }
}